Custos integration

machristie opened the pull request:
*Note: initially this PR aimed to add generic OIDC integration for any Keycloak instance. It has since been changed to add integration specific to Custos so as to simplify configuration and to keep ownership of the integration details with the Custos project.*

This pull request adds the ability to do OIDC authentication with Custos. The integration builds on the earlier python-social-auth integration support in Galaxy.

machristie committed:
Add implementation to OIDC backend config and routing

machristie committed:
Add implementation to OIDC backend config and routing

machristie committed:
Implemented Keycloak callback w. Galaxy user creation

machristie committed:
Implemented Keycloak callback w. Galaxy user creation

machristie committed:
Storing access_token, refresh_token in database

machristie committed:
Storing access_token, refresh_token in database

machristie committed:
Pass kc_idp_hint to redirect to identity provider login page directly

machristie committed:
Pass kc_idp_hint to redirect to identity provider login page directly

machristie committed:
Fix state parameter validation and store in cookie

machristie committed:
Fix state parameter validation and store in cookie

machristie committed:
Checking nonce; storing nonce and state in cookies instead of db

machristie committed:
Checking nonce; storing nonce and state in cookies instead of db

machristie committed:
Some small refactorings

machristie committed:
Some small refactorings

machristie committed:
Simplifying configuration by using .well-known config url

machristie committed:
Simplifying configuration by using .well-known config url

machristie committed:
Added unit tests, refactoring

machristie committed:
Added unit tests, refactoring

machristie committed:
Added tests around Galaxy user creation, keycloak_access_token records

machristie committed:
Added tests around Galaxy user creation, keycloak_access_token records

machristie committed:
Change db strategy to look up existing keycloak_access_token first

machristie committed:
Change db strategy to look up existing keycloak_access_token first

machristie commented:
**Note: these instructions no longer work. See below for how to test locally with Custos.**

Some notes on how to test locally with Keycloak:
1. Run Keycloak in Docker.
    ````
    docker run -d -p 9000:8080 -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin jboss/keycloak
    ```
2. Log into the Keycloak console at http://localhost:9000/auth with the username and password specified above.
3. Create a new client in the master realm: http://localhost:9000/auth/admin/master/console/#/realms/master/clients
    - specify a Client ID
    - click Save
    - set Access Type to confidential
    - set the Valid Redirect URI. The value will depend on what you named the "provider" in oidc_backends_config.xml. In this example, the provider is called "galaxy-auth": `http://localhost:8000/authnz/galaxy-auth/oidc/callback*`.
    - click Save
    - go to the Credentials tab and copy the client secret into oidc_backends_config.xml
4. Create a user to test with at http://localhost:9000/auth/admin/master/console/#/realms/master/users. You can set a non-temporary password after creating the user.
5. Set the well_known_oidc_config_uri in oidc_backends_config.xml. With this setup it would be: `http://localhost:9000/auth/realms/master/.well-known/openid-configuration`. The configuration will look something like this:
  ```xml
   <provider name="Galaxy-auth" implementation="OIDC">
        <client_id>galaxy-web</client_id>
        <client_secret>d2fac42f-8de0-4750-9e93-11ab7f5a9aed</client_secret>
        <redirect_uri>http://localhost:8000/authnz/galaxy-auth/oidc/callback</redirect_uri>
        <well_known_oidc_config_uri>http://localhost:9000/auth/realms/master/.well-known/openid-configuration</well_known_oidc_config_uri>
    </provider>
   ```

machristie commented:
@VJalili this is the Keycloak integration pull request. Let me know if you have any questions.

I think the only question I have now is whether `disconnect()` needs to be implemented and if so, what should it do.

VJalili commented on lib/galaxy/authnz/managers.py:
```diff
@@ -106,26 +111,48 @@ def _parse_google_config(self, config_xml):
+            rtv['idp_hint'] = config_xml.find('idp_hint').text
+        return rtv
+
+    def _unify_provider_implementation_names(self, provider, implementation):
```
Not all code paths return a value. For instance, if `v != provider` for all of enumerated items, the func returns `None`, which is an unexpected value for the caller of this function.

machristie replied:
Good catch. I'll change this to return `None, None` if there are no matches so it always returns a tuple.

VJalili replied:
Thanks. And probably also handling `None, None` and throwing proper exception.

VJalili commented on config/oidc_backends_config.xml.sample:
```diff
@@ -22,4 +22,20 @@
          login to Galaxy using their Google account.
         -->
     </provider>
+    <provider name="Keycloak-auth" implementation="Keycloak">
```
I guess it would be better to replace `keycloak-auth` with a provider name (e.g., `Google`) and set the `idp_hint` automatically accordingly.

machristie replied:
Well, the idp_hint has to be the alias of the identity provider, which could be whatever a user provides when configuring the identity provider in Keycloak, so we can't figure out the idp_hint from the name of the provider.

Or are you saying to flesh out the sample entry here as an example of how to add Google via Keycloak? If so, yeah that makes sense.

VJalili replied:
I guess I was confused with "how keycloak knows that it should redirect the user to Google, for instance"; because I don't see any reference to IdP, so I assumed it is the `idp_hint` specifying this backend (as you also . put it in the commend). I guess it could be intuitive (and consistent with PSA config) if it reads as `<provider name="google" implementation="Keycloak">`.

If you leave it as `<provider name="Keycloak-auth" implementation="Keycloak">` and then you want to implement login with X and Y idps, then we might end-up in configs such as the following: 

```xml
    <provider name="Keycloak-auth" implementation="Keycloak">
        ...
        <idp_hint>X</idp_hint>
    </provider>
    <provider name="Keycloak-auth" implementation="Keycloak">
        ...
        <idp_hint>Y</idp_hint>
    </provider>
```
or: 
```xml
    <provider name="Keycloak-auth" implementation="Keycloak">
        ...
        <idps>
                <idp_hint>X</idp_hint>
                <idp_hint>Y</idp_hint>
        </idp_hint>
    </provider>
```
but I guess the following would be more intuitive: 
```xml
    <provider name="X" implementation="Keycloak">
        ...
    </provider>
    <provider name="Y" implementation="Keycloak">
        ...
    </provider>
```

machristie replied:
Yes, that last one is what I had in mind, you would create two provider entries that could have different idp_hints.

VJalili commented on config/oidc_backends_config.xml.sample:
```diff
@@ -22,4 +22,20 @@
+        mykeycloak.org and the realm name is testrealm, then the URL is:
+        https://mykeycloak.org/auth/realms/testrealm/.well-known/openid-configuration
+        -->
+        <well_known_oidc_config_uri>https://.../auth/realms/.../.well-known/openid-configuration</well_known_oidc_config_uri>
```
Maybe it would be more intuitive to replace this attribute with something like `Keycloack_server` and `realm`, and generate `well_known_oidc_config_uri` in the backend.

machristie replied:
Yeah that would probably make it more user-friendly. I was trying to make it more forward compatible safe since Keycloak could decide at some point to change how they construct the .well-known url, but the risk of that seems small.  I'll change it to what you suggested.

VJalili commented on lib/galaxy/authnz/keycloak_authnz.py:
```diff
@@ -0,0 +1,147 @@
+            raise Exception("Nonce mismatch!")
+
+    def _load_well_known_oidc_config(self, well_known_uri):
+        return requests.get(well_known_uri).json()
```
This function raises the following unhandled exception.
`SSLError: HTTPSConnectionPool(host='localhost', port=9000): Max retries exceeded with url: /auth/realms/master/.well-known/openid-configuration (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'ssl3_get_record', 'wrong version number')],)",),))`

(I've set the `well_known_oidc_config_uri` to `https://localhost:9000/auth/realms/master/.well-known/openid-configuration`.)

machristie replied:
The URL should be "http" not "https".  See if that works.

VJalili replied:
Thats correct; however, it might be better to handle such an exception here, and also update the example well_known url.

VJalili replied:
Also with http I am getting `null` redirect uri, and checking terminal I see the following error:

```shell
  File "lib/galaxy/authnz/managers.py", line 233, in authenticate
    return True, "Redirecting to the `{}` identity provider for authentication".format(provider), backend.authenticate(trans)
  File "lib/galaxy/authnz/keycloak_authnz.py", line 44, in authenticate
    base_authorize_url, **extra_params)
  File "../.venv/lib/python2.7/site-packages/requests_oauthlib/oauth2_session.py", line 153, in authorization_url
    **kwargs), state
  File "../.venv/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/clients/web_application.py", line 86, in prepare_request_uri
    redirect_uri=redirect_uri, scope=scope, state=state, **kwargs)
  File "../.venv/lib/python2.7/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 71, in prepare_grant_uri
    raise InsecureTransportError()
InsecureTransportError: (insecure_transport) OAuth 2 MUST utilize https.
```
While this error makes a total sense, as it is per explicit requirements of OAuth 2.0 (and OIDC inherently) to TLS/SSL secure connection, but it can be tricky for local testing. Maybe this would better bond with the `VERIFY_SSL` setter `oidc_config.xml`. Those configurations should apply to all providers. So, while `VERIFY_SSL` is set to `False`, the backend should not throw exception/warning complaining of insecure connection.

machristie replied:
Did you create the environment variable: `export OAUTHLIB_INSECURE_TRANSPORT=1` ?

VJalili commented on lib/galaxy/authnz/managers.py:
```diff
@@ -106,26 +111,48 @@ def _parse_google_config(self, config_xml):
             log.debug(msg)
             return False, msg, None
 
+    def _get_authnz_backend_class(self, implementation):
```
`static` method?

machristie replied:
Yes

VJalili replied:
Would it make sense to make it static then? i.e., add `@staticmethod` decorator and remove `self`.

machristie replied:
Yes, I'll update it.

VJalili commented on lib/galaxy/authnz/managers.py:
```diff
@@ -188,17 +215,19 @@ def try_get_authz_config(sa_session, user_id, authz_id):
         return qres
 
-    def authenticate(self, provider, trans):
+    def authenticate(self, provider, implementation, trans):
```
If I use `provider = 'psa'` and `implementation = 'google'`, as a result of calling `authnz/psa/google/login` instead of `authnz/google/psa/login`, I get a `redirect_uri: null`. Maybe a more self-explanatory message would be better instead.

machristie replied:
Yeah, will do

VJalili commented:
@machristie Thank you for this integration. I added some inline comments. 

Regarding your notes on how to set this up for local testing: 
- I don't see ` Access Type` option at client creation, I see it in client settings tab after it is created.
- The `Credentials` tab appear after `Access type` is set to `confidential`. 

While testing this locally I am getting `Max retries exceeded with url` error, I added an inline comment. 

> I think the only question I have now is whether disconnect() needs to be implemented and if so, what should it do.

Since we're implementing this for psa, I guess for consistency between backends, it would be better to implement it for keycloak as well. 

Disconnect is the process of disassociating a Galaxy user and a third-party authnz. In other words, it is the process of removing any access and/or ID tokens of a user.

machristie commented:
Thanks @VJalili . I've updated the local testing instructions now.

I think it will be simple to implement the disconnect functionality. However, quick question, with the PSA implementation is there a way to connect an existing Galaxy user to a Google login? From what I could tell it looks like a PSA Google login would just create a new Galaxy user.

VJalili commented:
You can associate an existing Galaxy user with a third-party identity via PSA, and I guess this attribute should preserve with Keycloak as well. 

(if you're logged in with a Galaxy user account, and then you head to `/authnz/google/login` and complete the process, your Galaxy account will be associated with your Google account.)

One of the functions PSA leverages for this association is the following: 

https://github.com/galaxyproject/galaxy/blob/f954dc45fe94706f2ea04accc9fc92bcbcea892f/lib/galaxy/model/__init__.py#L5024-L5026

VJalili commented:
Consider the following function: 

https://github.com/galaxyproject/galaxy/blob/f954dc45fe94706f2ea04accc9fc92bcbcea892f/lib/galaxy/authnz/psa_authnz.py#L247-L250

This is currently implemented as part of PSA pipeline. And I guess a similar functionality should implement for Keycloak as well. Maybe it would be better to abstract this function, define a base that defines Galaxy's minimum requirement for both PSA and keycloak (e.g., `id token` must be available), then extend it as needed for keycloak and PSA.

machristie commented:
Going to try to keep track of some TODOs:
- [x] rename from Keycloak to OIDC, since the integration is generic and need not be dependent on Keycloak. This will require a couple small refactors to remove some Keycloak assumptions.
- [x] add support for associating an existing Galaxy user with an external identity.
- (to be implemented in separate pull request) when logging in through identity provider try to associate with an existing Galaxy user based on email address.
  - but need to send an email to the user to verify that they do in fact own this email address
  - see proposal below for details
- [x] implement `disconnect()`
- [x] Handle unknown provider/implementation in `_unify_provider_implementation_names`
- [x] better error message if provider/implementation are not found or mixed up in URL
- [x] simplify local testing, currently environment variable needs to be set
- [x] unit tests aren't running in travis

machristie committed:
Better handling of unrecognized provider/implementation pairs

machristie committed:
Better handling of unrecognized provider/implementation pairs

machristie committed:
Make _get_authnz_backend_class static

machristie committed:
Make _get_authnz_backend_class static

machristie committed:
Rename keycloak_authnz -> oidc_authnz

machristie committed:
Rename keycloak_authnz -> oidc_authnz

machristie committed:
Generalized extra auth url params and claim mapping

machristie committed:
Generalized extra auth url params and claim mapping

machristie committed:
Removed Keycloak from data model

machristie committed:
Removed Keycloak from data model

machristie committed:
Fixing py3 compatibility of oidc_authnz and tests

machristie committed:
Fixing py3 compatibility of oidc_authnz and tests

machristie committed:
Associate external id with currently logged in Galaxy user

machristie committed:
Associate external id with currently logged in Galaxy user

machristie committed:
Implemented `disconnect()` for OIDCAuthnz

machristie committed:
Implemented `disconnect()` for OIDCAuthnz

machristie committed:
Fixes for Google OIDC

Only include scopes in authorization request. When fetching a token,
Google will return different scopes than what are requested and oauthlib
raises a warning exception.

Check if a refresh_token exists in response instead of assuming it will be there.

machristie committed:
Fixes for Google OIDC

Only include scopes in authorization request. When fetching a token,
Google will return different scopes than what are requested and oauthlib
raises a warning exception.

Check if a refresh_token exists in response instead of assuming it will be there.

machristie commented:
Here's my proposal for email verification and associating a Galaxy user with an existing Galaxy user by email:
- I think oidc_authnz should respect the user activation setting in Galaxy. If it isn't enabled then 1) no email verification should take place and 2) no linking to existing Galaxy user accounts based on email address will take place
  - Discussed with @VJalili - if oidc is enabled then user activation will be required to be enabled as well. The reason is that we want to discourage the possibility of a user ending up with two Galaxy user accounts with the same email address which would happen if user activation is disabled and a user creates a Galaxy user account and then also logs in through OIDC with the same email address.  This does mean that to enable OIDC, either in production or locally for development, one will have to enable user activation [and anything else that user activation requires such as an smtp server and admin email](https://galaxyproject.org/admin/user-accounts/).
- if user activation is turned on, then the user will be sent an account activation link just as if they had registered with Galaxy.
- when the user verifies their email address and clicks on the activate URL, we can display a list of existing Galaxy user accounts that are also active and have the same email address. The user has the option of linking to one of those accounts or not. If they choose not to link then a new account is created or activated.
- we could also check if the `email_verified` claim is returned in the userinfo response. If so, oidc_authnz could bypass email verification and just activate the user directly and associate with the first existing Galaxy user account that has the same email address if any exist.

These changes will be made in a separate pull request.

machristie commented:
@VJalili I updated the email verification proposal above with what we discussed yesterday. Can you please review it and let me know if it still looks good and if I left anything out? Thanks!

machristie committed:
Automatically set env var when redirect_uri is http://localhost

machristie committed:
Automatically set env var when redirect_uri is http://localhost

machristie commented:
**Note: this PR no longer has generic OIDC support so this no longer works. To test this PR, see notes below on how to test the Custos integration.**

To test locally with Google OIDC, follow [the instructions here](https://galaxyproject.org/admin/authentication/config/) and use the following in `oidc_backends_config.xml`:
```xml
    <provider name="Google" implementation="OIDC">
        <client_id> ... </client_id>
        <client_secret> ... </client_secret>
        <redirect_uri>http://localhost:8080/authnz/google/callback</redirect_uri>
        <well_known_oidc_config_uri>https://accounts.google.com/.well-known/openid-configuration</well_known_oidc_config_uri>
        <userinfo_claim_mappings>
            <username>email</username>
        </userinfo_claim_mappings>
    </provider>
```

machristie commented:
@VJalili This PR is ready for a re-review whenever you have a chance. FYI, I'm out of the office until Monday.

bgruening commented:
@galaxybot test this

VJalili commented on config/oidc_backends_config.xml.sample:
```diff
@@ -22,4 +22,33 @@
          login to Galaxy using their Google account.
         -->
     </provider>
+    <provider name="Google" implementation="OIDC">
```
`OIDC` refers to all the providers, so I guess it would be better to name this `implementation="keyclock"` instead.

VJalili commented:
> * if oidc is enabled then user activation will be required to be enabled as well. The reason is that we want to discourage the possibility of a user ending up with two Galaxy user accounts with the same email address which would happen if user activation is disabled and a user creates a Galaxy user account and then also logs in through OIDC with the same email address.  This does mean that to enable OIDC, either in production or locally for development, one will have to enable user activation [and anything else that user activation requires such as an smtp server and admin email](https://galaxyproject.org/admin/user-accounts/).

Following our conversation, I agree we should require user activation enabled for OIDC to work in order to avoid user impersonation. However, I guess this maybe perceived an over-requirement to setup a development environment.

VJalili commented on lib/galaxy/model/mapping.py:
```diff
@@ -123,6 +123,22 @@
     Column('lifetime', Integer),
     Column('assoc_type', VARCHAR(64)))
 
+model.OIDCAccessToken.table = Table(
```
Since in addition to access token, we have id and refresh token among other values, I guess it would be better to have a more generic class name; e.g., `OIDCToken`.

VJalili replied:
Similarly, I guess table name can be more generic; e.g., `oidc_token`.

machristie replied:
Sounds good 👍

machristie committed:
Rename OIDCAccessToken -> OIDCToken

machristie committed:
Rename OIDCAccessToken -> OIDCToken

machristie committed:
Rename from generic "OIDC" to "Custos"

machristie committed:
Rename from generic "OIDC" to "Custos"

machristie committed:
Simplified configuration based on provider name and realm

machristie committed:
Simplified configuration based on provider name and realm

machristie committed:
Rolling back addition of `implementation` param

machristie committed:
Rolling back addition of `implementation` param

machristie committed:
Renumber custos_authnz_token migration

machristie committed:
Renumber custos_authnz_token migration

machristie committed:
Allow overriding idphint (defaults to 'cilogon')

machristie committed:
Allow overriding idphint (defaults to 'cilogon')

machristie committed:
Remove custos-dev provider

machristie committed:
Remove custos-dev provider

machristie committed:
Use InCommon CA cert to verify iam.scigap.org

machristie committed:
Use InCommon CA cert to verify iam.scigap.org

machristie commented:
How to test the Custos integration locally:

1. To test you'll use the production SciGaP Keycloak installation at https://iam.scigap.org/auth/. For that to work, however, you need to download the CA certificate that iam.scigap.org uses for its SSL certificate so that the SSL connection can be verified.
```
curl -O https://gist.githubusercontent.com/machristie/52a9f2715140df5fe88ea763dbd50b1a/raw/82aebbd7ea45df72e9f5757771e5272f8ea17498/incommon_rsa_server_ca.pem
mv incommon_rsa_server_ca.pem /path/to/galaxy/config/
```
2. Add the following `provider` config to oidc_backends_config.xml:
```xml
    <provider name="Custos">
        <url>https://iam.scigap.org/auth</url>
        <ca_bundle>/path/to/galaxy/config/incommon_rsa_server_ca.pem</ca_bundle>
        <client_id>local-galaxy</client_id>
        <client_secret>3b1e1256-d284-4cb6-b9bb-882f7bf8f06b</client_secret>
        <redirect_uri>http://localhost:8000/authnz/custos/callback</redirect_uri>
        <realm>galaxy-test</realm>
    </provider>
```
3. Start up Galaxy and go to http://localhost:8000/authnz/Custos/login. Copy and paste the redirect uri into the browser.
4. Sign in with CILogon.

machristie commented:
@VJalili This PR is ready for a re-review. See the previous comment for how to test the Custos integration.

VJalili commented on config/oidc_backends_config.xml.sample:
```diff
@@ -31,4 +31,16 @@
     </provider>
+
+    <provider name="Custos">
+        <client_id> ... </client_id>
```
I guess we should also have another required config for the Keycloak endpoint, named `keycloak_endpoint` for instance. Here you're using https://iam.scigap.org (see [this line](https://github.com/galaxyproject/galaxy/pull/7195/files#diff-74566421da44253f075336826614cf9cR194)) as the default endpoint; however, I guess it is better not to assume any defaults for this base endpoint and make it a required configuration.

VJalili commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -0,0 +1,212 @@
+        # TODO: Look up this URL from a Python library
+        if provider == 'custos':
+            self.config['ca_bundle'] = CA_CERTFILE
+            return "https://iam.scigap.org/auth/realms/{}/.well-known/openid-configuration".format(realm)
```
I guess the endpoint part of the URL (i.e., https://iam.scigap.org) can be defined in the config. (See my comment on the config sample.)

VJalili commented on lib/galaxy/authnz/incommon_rsa_server_ca.pem:
```diff
@@ -0,0 +1,68 @@
+-----BEGIN CERTIFICATE-----
```
I guess this file is committed accidentally.

VJalili commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -0,0 +1,212 @@
+STATE_COOKIE_NAME = 'custos-state'
+NONCE_COOKIE_NAME = 'custos-nonce'
+BASE_DIR = os.path.dirname(os.path.abspath(__file__))
+CA_CERTFILE = os.path.join(BASE_DIR, "incommon_rsa_server_ca.pem")
```
I guess Galaxy's `config` folder could be a better place for storing/retrieving RSA certificates; because with code, the certificates are expected to be located at `galaxy/lib/authnz`, hence making config and code colocated, which is an uncommon setup in Galaxy.

VJalili commented on config/oidc_backends_config.xml.sample:
```diff
@@ -31,4 +31,16 @@
+
+        <realm> ... </realm>
+        <!-- (Optional) Override the default Custos well-known URL to point to a different instance -->
+        <!-- <well_known_oidc_config_uri>https://.../.well-known/openid-configuration</well_known_oidc_config_uri> -->
```
Maybe we can expect admins to provide only the `.well-known` part of the URL for this setter (e.g., `.well-known/openid-configuration</well_known_oidc_config_uri`), and construct the full URL in the `custos_authnz` backend by concatenating this setter with the base URL (i.e., `https://.../`; see `keycloak_endpoint` in my previous comment).

In other words, we can assume the URLs are made as `<KEYCLOAK_ENDPOINT> + <.WELL-KNOWN>`, where `KEYCLOAK_ENDPOINT`, set using `keycloak_endpoint` setter is mandatory, and `WELL-KNOWN` defined by the `well_known_oidc_config_uri` setter is optional with default values.

VJalili commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -0,0 +1,212 @@
+                and os.environ.get("OAUTHLIB_INSECURE_TRANSPORT", None) != "1"):
+            log.warn("Setting OAUTHLIB_INSECURE_TRANSPORT to '1' to "
+                     "allow plain HTTP (non-SSL) callback")
+            os.environ['OAUTHLIB_INSECURE_TRANSPORT'] = "1"
```
Do you think it would make any sense to set `OAUTHLIB_INSECURE_TRANSPORT` leveraging the [VERIFY_SSL](https://github.com/galaxyproject/galaxy/blob/31a5fc24c1ce1c9d902535eb06292bedfdb6efd4/config/oidc_config.xml.sample#L16) config?

machristie replied:
No because VERIFY_SSL is used to control whether to verify SSL of the authorization server whereas OAUTHLIB_INSECURE_TRANSPORT is used to enable insecure callback on the client server.

VJalili commented on lib/galaxy/authnz/custos_authnz.py:
```diff
@@ -0,0 +1,212 @@
+    def _get_current_user(self, trans):
+        return trans.user if trans.user else None
+
+    def _create_user(self, sa_session, username, email):
```
Do you think using [this create_user](https://github.com/galaxyproject/galaxy/blob/31a5fc24c1ce1c9d902535eb06292bedfdb6efd4/lib/galaxy/model/__init__.py#L5082-L5089) method instead would make any sense?

VJalili commented on lib/galaxy/model/__init__.py:
```diff
@@ -5099,6 +5099,22 @@ def create_social_auth(cls, user, uid, provider):
+        self.refresh_token = refresh_token
+        self.expiration_time = expiration_time
+        self.refresh_expiration_time = refresh_expiration_time
+        self.raw_token = raw_token
```
I guess now that you're storing `access_token`, `id_token`, `refresh_token`, `expiration_time`, and etc., which are all part of the `raw_token`, you would not need to persist the `raw_token` it self.

VJalili commented on test/unit/authnz/test_custos_authnz.py:
```diff
@@ -0,0 +1,474 @@
+
+    def setUp(self):
+        self.orig_requests_get = requests.get
+        requests.get = self.mockRequest("https://iam.scigap.org/auth/realms/test-realm/.well-known/openid-configuration", {
```
My comment about the config setter also applies to this endpoint.

VJalili committed:
Add URL config key.

VJalili committed:
Replace scigap endpoint with a URL set in the config for custos backend.

VJalili committed:
Update custos unit test.

machristie committed:
Merge branch 'keycloak-integration' into patch-5

machristie committed:
Merge pull request #2 from VJalili/patch-5

Use a base URI set in config for Custos Authnz backend.

machristie committed:
Fixing tests

machristie committed:
Remove CA cert, add ca_bundle config option

machristie committed:
Removed raw_token column, bumped migration number

machristie commented:
Hi @VJalili . I've integrated your changes (thanks!) and I've also removed the raw_token column and made the ca_bundle a config option and removed the CA certificate from the PR. Can you please re-review when you have a chance?

I'll update the instructions on how to test locally.

VJalili commented:
@machristie LGTM, +1, Thanks. 

In a different PR, I think we should work on the following function, and make sure our tests are not tied to that specific endpoint.

```python
def _get_base_idp_url(self):
    # it would be ideal is we can use a URI as the following:
    # https://test_base_uri/auth
    return 'https://iam.scigap.org/auth'
```

afgane commented:
Nice work @machristie - thank you a bunch for pushing this through!

machristie commented:
Thanks @afgane and thanks to @VJalili for his patience and dedication to making this PR a success!
